Fix checking of non handled data for Values format

azat opened the pull request:
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix checking of non handled data for Values format

PeekableReadBuffer::hasUnreadData() does not checks the underlying buffer, and so it simply ignore some issues, like:

    INSERT INTO test_01179_str values ('foo'); ('bar')

Cc: @tavplubix 
Cc: @zvonand

_P.S. this is actually not a bug fix at all, but I've just wanted to show this change in the changelog, since it can break some obscure usage. I was thinking between `bug fix` and `backward incompatible`, but later sounds too serious for this change, though maybe `bacward incompatible` will be better..._

azat committed:
Fix checking of non handled data for Values format

PeekableReadBuffer::hasUnreadData() does not checks the underlying
buffer, and so it simply ignore some issues, like:

    INSERT INTO test_01179_str values ('foo'); ('bar')

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

robot-ch-test-poll4 commented:
<!-- automatic status comment for PR #55527 from azat/ClickHouse:values-eof-check-fix -->
*This is an automated comment for commit f35a15efff96905d5d1d0c5c19b7e01c1caa0734 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/55527/f35a15efff96905d5d1d0c5c19b7e01c1caa0734/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6486752067">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/55527/f35a15efff96905d5d1d0c5c19b7e01c1caa0734/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

azat committed:
Move comment into correct place

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Add ability to ignore data after semicolon in Values format

This is required for client, to handle comments in multiquery mode.

v0: separate context for input format
v2: cannot use separate context since params and stuff are changed in global context
v3: do not sent this setting to the server (breaks queries for readonly profiles)
Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented:
So it cannot be done that simple, need to allow extra data after semicolon for client/local multiquery mode, I've added separate setting for this `input_format_values_allow_data_after_semicolon` which is automatically enabled in case `--multiquery`.

azat committed:
Update 00626_replace_partition_from_table_zookeeper to avoid using comments after semicolon

It does not have multiquery, so better to use shell comments

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

tavplubix commented on programs/client/Client.cpp:
```diff
@@ -1264,6 +1264,15 @@ void Client::processConfig()
+    {
+        Settings settings = global_context->getSettings();
+        settings.input_format_values_allow_data_after_semicolon = true;
+        /// Do not send it to the server
+        settings.input_format_values_allow_data_after_semicolon.changed = false;
```
That's too tricky, but maybe it's ok

azat replied:
The same is used for distributed queries, so I think it is OK:

```
src/Interpreters/ClusterProxy/executeQuery.cpp:        new_settings.max_concurrent_queries_for_user.changed = false;
src/Interpreters/ClusterProxy/executeQuery.cpp:        new_settings.max_memory_usage_for_user.changed = false;
src/Interpreters/ClusterProxy/executeQuery.cpp:            new_settings.force_optimize_skip_unused_shards.changed = false;
src/Interpreters/ClusterProxy/executeQuery.cpp:            new_settings.optimize_skip_unused_shards.changed = false;
src/Interpreters/ClusterProxy/executeQuery.cpp:        new_settings.offset.changed = false;
src/Interpreters/ClusterProxy/executeQuery.cpp:        new_settings.limit.changed = false;
```

azat commented:
>Stateless tests (aarch64) — fail: 1, passed: 5706, skipped: 28

- `02263_lazy_mark_load` - this test is [flaky](https://play.clickhouse.com/play?user=play#U0VMRUNUIGNoZWNrX25hbWUsIHRlc3RfbmFtZSwgcmVwb3J0X3VybApGUk9NIGNoZWNrcwpXSEVSRSB0ZXN0X25hbWUgbGlrZSAnJTAyMjYzX2xhenlfbWFya19sb2FkJScKICAgIEFORCBjaGVja19zdGFydF90aW1lID49IG5vdygpIC0gSU5URVJWQUwgMjQgSE9VUgovLyAgICBBTkQgcHVsbF9yZXF1ZXN0X251bWJlciA9IDAKICAgIEFORCB0ZXN0X3N0YXR1cyAhPSAnU0tJUFBFRCcKICAgIEFORCB0ZXN0X3N0YXR1cyBMSUtFICdGJScKICAgIEFORCBjaGVja19zdGF0dXMgIT0gJ3N1Y2Nlc3MnCk9SREVSIEJZIGNoZWNrX25hbWUsIHRlc3RfbmFtZSwgY2hlY2tfc3RhcnRfdGltZQ==)

>Stateless tests (tsan) [5/5] — Tests are not finished, fail: 2, passed: 236, skipped: 1

- `02479_race_condition_between_insert_and_droppin_mv`
- `02421_truncate_isolation_with_mutations`

So as this two.
